plugins/GranularPitchShifter/GranularPitchShifter.cpp

```diff
@@ -0,0 +1,269 @@
+	Effect(&granularpitchshifter_plugin_descriptor, parent, key),
+	m_granularpitchshifterControls(this),
+	m_prefilter({PrefilterLowpass(), PrefilterLowpass()})
+{
+	autoQuitModel()->setValue(autoQuitModel()->maxValue());
+	
+	changeSampleRate();
+}
+
+
+bool GranularPitchShifterEffect::processAudioBuffer(sampleFrame* buf, const fpp_t frames)
+{
+	if (!isEnabled() || !isRunning()) { return false; }
+
+	const float d = dryLevel();
+	const float w = wetLevel();
+	
+	const ValueBuffer * pitchBuf = m_granularpitchshifterControls.m_pitchModel.valueBuffer();
+	const ValueBuffer * pitchSpreadBuf = m_granularpitchshifterControls.m_pitchSpreadModel.valueBuffer();
+
+	const float size = m_granularpitchshifterControls.m_sizeModel.value();
+	const float shape = m_granularpitchshifterControls.m_shapeModel.value();
+	const float jitter = m_granularpitchshifterControls.m_jitterModel.value();
+	const float twitch = m_granularpitchshifterControls.m_twitchModel.value();
+	const float spray = m_granularpitchshifterControls.m_sprayModel.value();
+	const float spraySpread = m_granularpitchshifterControls.m_spraySpreadModel.value();
+	const float density = m_granularpitchshifterControls.m_densityModel.value();
+	const float glide = m_granularpitchshifterControls.m_glideModel.value();
+	const int minLatency = m_granularpitchshifterControls.m_minLatencyModel.value() * m_sampleRate;
+	const float densityInvRoot = std::sqrt(1.f / density);
+	const float feedback = m_granularpitchshifterControls.m_feedbackModel.value() * densityInvRoot;
+	const float fadeLength = 1.f / m_granularpitchshifterControls.m_fadeLengthModel.value();
+	const bool prefilter = m_granularpitchshifterControls.m_prefilterModel.value();
+	
+	if (glide != m_oldGlide)
+	{
+		m_oldGlide = glide;
+		m_glideCoef = std::exp(-1 / (glide * m_sampleRate));
+	}
+	
+	const float shapeK = cosWindowApproxK(shape);
+	const int sizeSamples = m_sampleRate / size;
+	const float waitMult = sizeSamples / (density * 2);
+
+	for (fpp_t f = 0; f < frames; ++f)
+	{
+		const double pitch = pitchBuf ? pitchBuf->value(f) : m_granularpitchshifterControls.m_pitchModel.value();
+		const double pitchSpread = (pitchSpreadBuf ? pitchSpreadBuf->value(f) : m_granularpitchshifterControls.m_pitchSpreadModel.value()) * 0.5f;
+		
+		// interpolate pitch depending on glide
+		for (int i = 0; i < 2; ++i)
+		{
+			double targetVal = pitch + pitchSpread * (i ? 1. : -1.);
+			
+			if (targetVal == m_truePitch[i]) { continue; }
+			m_updatePitches = true;
+			
+			m_truePitch[i] = m_glideCoef * m_truePitch[i] + (1. - m_glideCoef) * targetVal;
+			// we crudely lock the pitch to the target value once it gets close enough, so we can save on CPU
+			if (std::abs(targetVal - m_truePitch[i]) < GPS_GLIDE_SNAG_RADIUS) { m_truePitch[i] = targetVal; }
+		}
+		
+		// this stuff is computationally expensive, so we should only do it when necessary
+		if (m_updatePitches)
```

Comment:
I believe pulling out this "computationally expensive" if block to a seperate function might please codefactor

Reply:
We had a discussion about CodeFactor's complex method complaints on Discord with past PRs and have concluded that it's best to ignore it.  I don't think I'll ever be releasing a plugin that satisfies these.  It's a measurement of Cyclomatic complexity: https://en.wikipedia.org/wiki/Cyclomatic_complexity

Leaving this part in this function does not hinder readability in any way, and pulling it out of that function would result in 44100 unnecessary function calls per second, unless it's inlined in which case it would have to be placed in the header file, which _would_ hinder readability.

Reply:
I really want to just disable codefactor, we ignore it anyway

Reply:
I've told codefactor to ignore this issue.  If this becomes a burden, we can ignore the entire file, or disable it project-wide.  Resolving.